### PR TITLE
Parsing tests for CSS UI Level 3

### PR DESCRIPTION
**Pull request by ewilligers**
The tests expose a number of bugs:

https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/16854772/
https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/16900961/

https://bugzilla.mozilla.org/show_bug.cgi?id=1453258 [fixed]
https://bugzilla.mozilla.org/show_bug.cgi?id=1457652 [fixed]

https://bugs.chromium.org/p/chromium/issues/detail?id=708197 [fixed]
https://bugs.chromium.org/p/chromium/issues/detail?id=831550

### CODE REVIEW

**Review by emilio — APPROVED** [review]

File: css/css-ui/parsing/outline-valid.html [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+<script src="/resources/testharnessreport.js"></script>
+<script src="resources/parsing-testcommon.js"></script>
+</head>
+<body>
+<script>
+test_valid_value("outline", "invert");
+test_valid_value("outline", "rgba(10, 20, 30, 0.4)");
+
+test_valid_value("outline", "auto");
+test_valid_value("outline", "none", ["none", "invert"]); // Edge serializes as "invert"
```
[the remark this answers is not in the document]
Comment (emilio) [reply css/css-ui/parsing/outline-valid.html]: Or if it does, please leave a comment with the reasoning from the spec text.

[the remark this answers is not in the document]
Comment (ewilligers) [reply css/css-ui/parsing/outline-valid.html]: I have now omitted "invert" as an accepted serialization.

**Review by frivoal — CHANGES_REQUESTED** [review]
Initially, I thought that this looks good (modulo the few minor fixes I've requested).

That said, separately from evaluating the correctness of these tests, I have a little bit of concern about the effects of testing syntax only, as this series of tests does.

Quite a lot of this (most of the *-valid.html) tests are redundant with existing tests, as the existing ones couldn't pass if valid syntax wasn't supported (exceptions for things like checking calc(), which may not have full coverage in the existing test suite). But a fair amount of the existing tests are also manual (and hard/impossible to convert), and there's value in automated checking.

However, in the past, browsers have on occasion added parsing support for some property/values without actually adding the corresponding behavior:
* implement syntax first, behavior later
* make a syntax improvement that affects several properties, implement the behavior only in some of them (this is a subcase of the previous point)
* implement both syntax and behavior, maintain the parsing well but let the behavior bitrot and eventually fall apart

That sort of things makes it hard for web developers to provide fallbacks, either using the cascade or `@supports`, since they cannot reliably detect whether the feature is actually supported.

Because of this, I am a little concerned about syntax-only tests, which make it looks like implementing syntax only is an improvement, while it is actually a source of problems.

Those that test for invalid syntax are fine, as this concern does not apply, and those that are known to catch actual browser bugs are especially useful.

But for those that test for valid syntax, I am a little less sure:
* if the effects of the property/value can be automatically tested, aren't those tests enough and more safe?
* if the effects of the property/value cannot be automatically tested, how do we avoid turning this into a temptation to support syntax without behavior?

Here's one example of how things could go wrong here: a browser who has no interest in supporting actual inverted color outlines implements support for parsing `outline:invert` without the behavior, just to make a test go green. The acceptable behaviors (parse and do it, or fail to parse) are already covered in outline-006.html, so maybe that test provides the necessary safety. At the same time, given that that test exists, do we need parsing-only tests for `invert`?

I wonder what's the best approach. Here are a few suggestions:
1. the parts of of *-valid tests that check that things are not serialized to some known-to-be incorrect value are useful, (e.g. `test_valid_value("outline", "none"); // Edge serializes as "invert"`, or the tests that mozilla fails due to incorrect ordering or not using the shortest serialization), but they could be changed from "assert equals the correct thing" to "assert not equal the known-to-be-wrong thing".
2. make reftests for the parts of *-valid tests that can and that that are not already covered in the test suite (e.g. calc in outline)
3. drop the rest of -valid tests

(we should probably do 2. anyway, and possibly also 1., even if we don't plan to do 3.)

I know this isn't very satisfying, because this series of test is nice on its own, and some parts that would be dropped are only covered in manual tests. But the concern remains.

We could add warnings in the tests ("if this fails, don't fix it unless you also make sure that &lt;a href="...">other test&lt:/a> also passes"), but I am far from sure this would be effective.

Or maybe we just trust that reason will prevail and browser developers will not do the bad things I'm worried about. But as it has happened in the past, I am a little nervous about providing incentives (in the form of failing tests) for the unfortunate behavior. 

Or maybe I'm just paranoid, and this is fine. Feedback appreciated.

CCing a few CSS tests gurus for opinions: @fantasai @gsnedders @dbaron @Ms2ger

**Comment by ewilligers** [comment]
@fantasai has made a similar [comment](https://github.com/w3c/csswg-drafts/issues/1329#issuecomment-360940438
) about the dangers of parsing tests.

However, these parsing/serialization tests have been finding bugs in all implementations and have prompted fixes in a couple. This suggests their value.

File: css/css-ui/parsing/outline-valid.html [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+<link rel="author" title="Eric Willigers" href="mailto:ericwilligers@chromium.org">
+<link rel="help" href="https://drafts.csswg.org/css-ui-3/#outline">
+<meta name="assert" content="outline supports the full grammar '<outline-color> || <outline> || <outline>'.">
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script src="resources/parsing-testcommon.js"></script>
+</head>
+<body>
+<script>
+test_valid_value("outline", "invert");
```
[the remark this answers is not in the document]
Comment (ewilligers) [reply css/css-ui/parsing/outline-valid.html]: done

File: css/css-ui/parsing/outline-valid.html [hunk 3]
```diff
@@ -0,0 +1,42 @@
+<!DOCTYPE html>
+<html>
+<head>
+<meta charset="utf-8">
+<title>CSS UI Level 3: parsing outline with valid values</title>
+<link rel="author" title="Eric Willigers" href="mailto:ericwilligers@chromium.org">
+<link rel="help" href="https://drafts.csswg.org/css-ui-3/#outline">
+<meta name="assert" content="outline supports the full grammar '<outline-color> || <outline> || <outline>'.">
```
[the remark this answers is not in the document]
Comment (ewilligers) [reply css/css-ui/parsing/outline-valid.html]: done, but that applies applies to nearly all the -valid tests.

File: css/css-ui/parsing/outline-color-valid.html [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+<link rel="help" href="https://drafts.csswg.org/css-ui-3/#outline-color">
+<meta name="assert" content="outline-color supports the full grammar '<color> | invert'.">
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script src="resources/parsing-testcommon.js"></script>
+</head>
+<body>
+<script>
+test_valid_value("outline-color", "rgba(10, 20, 30, 0.4)");
+test_valid_value("outline-color", "invert");
```
[the remark this answers is not in the document]
Comment (ewilligers) [reply css/css-ui/parsing/outline-color-valid.html]: done

**Comment by emilio** [comment]
I don't think any browser implementer would seriously consider shipping syntax first without any behavior, is there any precedent for that?

**Comment by emilio** [comment]
(Apart from WebKit shipping parsing of `display: contents` but only honoring it on `<slot>`s, but that had a lot of other implications).

**Comment by gsnedders** [comment]
@frivoal 

> Because of this, I am a little concerned about syntax-only tests, which make it looks like implementing syntax only is an improvement, while it is actually a source of problems.

Note we essentially have the same issue with everything using idlharness.js: you get thousands of tests passing if you dump the IDL file in your tree and your binding generation is correct.

It's never been the case that number of tests passed has ever been meaningful (because we have both extremes in various places: ones where you can have 99.9% of the spec correct and fail almost everything, and ones where you can have 0.1% of the spec correct and pass almost everything), and I don't think we should start worrying about it now.

Ideally, we'd have something analogue to idlharness.js for CSS (where you give the property value definition and it generates however many tests for the syntax being supported), which is #9113.

We know these tests catch issues, so we absolutely should have them.

**Comment by FremyCompany** [comment]
> how do we avoid turning this into a temptation to support syntax without behavior?
I think this is a valid concern in theory, but not really an issue in practice. Speaking for Edge here, tests aren't ever going to be the reason we would do this. There is nothing worse than gaining a few points on a test suite with a fake implementation that later causes you to have a bug on a real website. No user is ever looking at test suites, users look at real sites. 

Now, it's totally possible that Edge would ship syntax even if it doesn't support its behavior, for instance if we face real websites being broken (we have shipped `getMatchedCssRules` as a function returning null because some website called it and crashed if the function didn't exist but never actually used the results for anything, and from that day on we have been working with Chrome to have them remove this feature, and they finally did, so we will be able to remove this shim from our code too, yeah!). But as horrible as these kind of stories are, I don't think anyone here would consider passing a wpt test important enough to actually ship a broken feature. Broken behaviors can break real sites, we don't want to take that risk.

FWIW though, I was doing exactly this right now in a prototype pull request because Edge does not parse `cursor: url(...) <number> <number>, fallback" because we don't support cursors that are not of the `*.cur` format which doesn't need these numbers; however since we don't parse it, we don't even get the fallback value in Edge, so it's worth parsing these numbers even if we don't use them; these tests made me realize I hadn't thought about parsing `calc()` for a number so I'd say they were actually useful.

**Comment by frivoal** [comment]
side discussion
@FremyCompany Your case with cursor and numbers is interesting. For non .cur cursors, this is an improvements and gives a better fallback. For .cur, in theory the two numbers allow authors to override the embedded hotspot, and your fix prevents them from being able to detect that this won't work. But I guess this is sufficiently rare that it's worth doing anyway. I'd be nice if you fully supported the explicit hotspot coordinate, though. Also, don't you support the .ico format in addition to .cur, and this doesn't have built-in coordinates?

**Comment by frivoal** [comment]
> I think this is a valid concern in theory, but not really an issue in practice.

I am not sure (I mean, literally, I don't mean that I am sure of the opposite).

Multiple browsers have in the past made parsing only implementations (and I am not talking about the deliberate kind of things like parsing the hotspot of the cursor, which can be argued for).

Multiple browsers have, on occasion, made races out of completing a test suite or having the highest pass percentage or score.

Should people know better? Totally. Do people still fall for it on occasion? I don't know for sure, but I'd guess so. Maybe junior developers. Maybe not-so-junior developers under pressure from a marketing team.

I wouldn't go as far as to say parsing-only tests should never exist, but I don't think they should be routine, and we need to weight the pros and cons each time.

Here, I'm thinking the ones that correspond to things that can only be tested manually are probably good ideas. The rest, I am less sure.

**Comment by csnardi** [comment]
I echo others who have already stated that these tests are useful. Similar tests have exposed bugs in implementations such as https://crbug.com/776208, where Chrome allowed syntax that was explicitly rejected by the spec, and areas where serialization differs between browsers (due to implementations or the spec being unclear). While testing that output via content is important, testing interoperability of syntax and serialization through CSSOM/Typed OM is additionally important, and thus both types of tests are needed.

**Review by frivoal — APPROVED** [review]
Alright, given that lots of people find these useful, and given the unique situation of css-ui (many manual tests), I'll stop complaining about these :) Approved.

I'll redirect my energy about this concern to https://github.com/w3c/web-platform-tests/issues/9113

**Pull request merged**